FirebaseInstallations/Source/Library/SecureStorage/FIRInstallationsKeychainUtils.m

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "FIRInstallationsKeychainUtils.h"
+
+#import "FIRInstallationsErrorUtil.h"
+
+@implementation FIRInstallationsKeychainUtils
+
++ (nullable NSData *)getItemWithQuery:(NSDictionary *)query
+                                error:(NSError *_Nullable *_Nullable)outError {
+  NSMutableDictionary *mutableQuery = [query mutableCopy];
+
+  mutableQuery[(__bridge id)kSecReturnData] = @YES;
+  mutableQuery[(__bridge id)kSecMatchLimit] = (__bridge id)kSecMatchLimitOne;
+
+  CFDataRef result = NULL;
+  OSStatus status =
+      SecItemCopyMatching((__bridge CFDictionaryRef)mutableQuery, (CFTypeRef *)&result);
+
+  if (status == errSecSuccess && result != NULL) {
+    if (outError) {
+      *outError = nil;
+    }
+
+    return (__bridge_transfer NSData *)result;
+  }
+
+  if (status == errSecItemNotFound) {
+    if (outError) {
+      *outError = nil;
+    }
+  } else {
+    if (outError) {
+      *outError = [FIRInstallationsErrorUtil keychainErrorWithFunction:@"SecItemCopyMatching"
+                                                                status:status];
+    }
+  }
+  return nil;
+}
+
++ (BOOL)setItem:(NSData *)item
```

Comment:
Are the write the operations off mian thread?  Might consider doing that later as we got issue before when keychain operation stuck when it's running on main thread.

Reply:
This method is just a sync helper, so it is up to caller what thread to perform the operation on. This particular method is called from [here](https://github.com/firebase/firebase-ios-sdk/pull/4114/files#diff-af0497ab8d81ffca80c9d5cb3c98cc8dR99) from a keychain dedicated queue.